endpoints/put.go

```diff
@@ -32,113 +31,199 @@ func NewPutHandler(backend backends.Backend, maxNumValues int, allowKeys bool) f
 
-		err = json.Unmarshal(body, put)
+		backend.put = putAnyRequestPool.Get().(*PutRequest)
+		defer putAnyRequestPool.Put(backend.put)
+
+		err = json.Unmarshal(body, backend.put)
 		if err != nil {
 			http.Error(w, "Request body "+string(body)+" is not valid JSON.", http.StatusBadRequest)
 			return
 		}
 
-		if len(put.Puts) > maxNumValues {
-			http.Error(w, fmt.Sprintf("More keys than allowed: %d", maxNumValues), http.StatusBadRequest)
+		// Get a response object from the resource pool that we'll fill with processed info
+		backend.resp = putResponsePool.Get().(*PutResponse)
+		backend.resp.Responses = make([]PutResponseObject, len(backend.put.Puts))
+		backend.resp.toCacheStrings = make([]string, len(backend.put.Puts))
+		defer putResponsePool.Put(backend.resp)
+
+		var cancel func()
+		backend.ctx, cancel = context.WithTimeout(context.Background(), 500*time.Millisecond)
+		defer cancel()
+
+		exitData := &exitInfo{errMsg: "", status: http.StatusOK}
+
+		if len(backend.put.Puts) == 0 {
+			exitData.errMsg = "No keys were sent in backend request"
+			exitData.status = http.StatusBadRequest
+		} else if len(backend.put.Puts) > maxNumValues {
+			exitData.errMsg = fmt.Sprintf("More keys than allowed: %d", maxNumValues)
+			exitData.status = http.StatusBadRequest
+		} else { // 1 <= len(backend.put.Puts) < maxNumValues
+			// Processonly if all requests come error free
+			validateAndEncode(backend.put, backend.resp, exitData)
+			if exitData.status == http.StatusOK {
+				if len(backend.put.Puts) == 1 {
+					callBackendGet(backend, 0)
+					callBackendPut(backend, exitData, 0)
+				} else {
+					// Run process in parallel
+					indexChannel := make(chan int)
+					done := make(chan bool)
+
+					go func() {
+						for i, _ := range backend.put.Puts {
+							indexChannel <- i
+						}
+						close(indexChannel)
+					}()
+
+					go callBackendInParallel(backend, exitData, indexChannel, done)
+					<-done
+				}
+			}
+		}
+
+		if exitData.status != http.StatusOK {
+			http.Error(w, exitData.errMsg, exitData.status)
 			return
 		}
 
-		resps := putResponsePool.Get().(*PutResponse)
-		resps.Responses = make([]PutResponseObject, len(put.Puts))
-		defer putResponsePool.Put(resps)
+		bytes, err := json.Marshal(backend.resp)
+		if err != nil {
+			http.Error(w, "Failed to serialize UUIDs into JSON.", http.StatusInternalServerError)
+			return
+		}
 
-		for i, p := range put.Puts {
-			if len(p.Value) == 0 {
-				http.Error(w, "Missing value.", http.StatusBadRequest)
-				return
-			}
-			if p.TTLSeconds < 0 {
-				http.Error(w, fmt.Sprintf("request.puts[%d].ttlseconds must not be negative.", p.TTLSeconds), http.StatusBadRequest)
+		/* Handles POST */
+		w.Header().Set("Content-Type", "application/json")
+		w.Write(bytes)
+	}
+}
+
+func validateAndEncode(puts *PutRequest, resp *PutResponse, exit *exitInfo) {
+	for index, p := range puts.Puts {
+		var toCache string
+
+		if len(p.Value) == 0 {
+			exit.errMsg = "Missing value."
```

Comment:
Consider returning new exitInfo objects instead of alerting an existing one. You can avoid memory duplication by doing the same in the other `if` branchea.

Reply:
`http.StatusOK` gets assigned to the `exitData` object in line 63 and we test for the status to keep being OK in lines  74 and 92.  Isn't it beter to pass the `exitData` pointer to both `validateAndEncode(backend *backendCallObject, exit *exitInfo)` and `callBackendPut(backend *backendCallObject, exit *exitInfo, index int)`? I suppose the pointer approach will create a single memory allocation in `NewPutHandler`'s stack and by passing by reference to `validateAndEncode(backend *backendCallObject, exit *exitInfo)` and `callBackendPut(backend *backendCallObject, exit *exitInfo, index int)`, aren't we only creating `int`s in their respective stacks as opposed to additional `exitInfo` objects in their respective stacks. I can change it if you want though.